libcaf_net/test/datagram_transport.cpp

```diff
@@ -0,0 +1,261 @@
+ ******************************************************************************/
+
+#define CAF_SUITE datagram_transport
+
+#include "caf/net/datagram_transport.hpp"
+
+#include "caf/net/test/host_fixture.hpp"
+#include "caf/test/dsl.hpp"
+
+#include "caf/byte.hpp"
+#include "caf/make_actor.hpp"
+#include "caf/net/actor_proxy_impl.hpp"
+#include "caf/net/endpoint_manager.hpp"
+#include "caf/net/endpoint_manager_impl.hpp"
+#include "caf/net/make_endpoint_manager.hpp"
+#include "caf/net/multiplexer.hpp"
+#include "caf/net/socket_guard.hpp"
+#include "caf/net/udp_datagram_socket.hpp"
+#include "caf/serializer_impl.hpp"
+#include "caf/span.hpp"
+
+using namespace caf;
+using namespace caf::net;
+
+namespace {
+
+constexpr string_view hello_manager = "hello manager!";
+
+class dummy_application_factory;
+
+struct fixture : test_coordinator_fixture<>, host_fixture {
+  using buffer_type = std::vector<byte>;
+
+  using buffer_ptr = std::shared_ptr<buffer_type>;
+
+  fixture() : shared_buf(std::make_shared<buffer_type>(1024)) {
+    mpx = std::make_shared<multiplexer>();
+    if (auto err = mpx->init())
+      CAF_FAIL("mpx->init failed: " << sys.render(err));
+    mpx->set_thread_id();
+    CAF_CHECK_EQUAL(mpx->num_socket_managers(), 1u);
+    if (auto err = parse("127.0.0.1:0", ep))
+      CAF_FAIL("parse returned an error: " << sys.render(err));
+    auto send_pair = unbox(make_udp_datagram_socket(ep));
+    send_socket = send_pair.first;
+    auto receive_pair = unbox(make_udp_datagram_socket(ep));
+    recv_socket = receive_pair.first;
+    ep.port(htons(receive_pair.second));
+    CAF_MESSAGE("sending message to " << CAF_ARG(ep));
+    if (auto err = nonblocking(recv_socket, true))
+      CAF_FAIL("nonblocking() returned an error: " << err);
+  }
+
+  ~fixture() {
+    close(send_socket);
+    close(recv_socket);
+  }
+
+  bool handle_io_event() override {
+    return mpx->poll_once(false);
+  }
+
+  error read_from_socket(udp_datagram_socket sock, buffer_type& buf) {
+    uint8_t receive_attempts = 0;
+    variant<std::pair<size_t, ip_endpoint>, sec> read_ret;
+    do {
+      read_ret = read(sock, make_span(buf));
+      if (auto read_res = get_if<std::pair<size_t, ip_endpoint>>(&read_ret)) {
+        buf.resize(read_res->first);
+      } else if (get<sec>(read_ret) != sec::unavailable_or_would_block) {
+        return make_error(get<sec>(read_ret), "read failed");
+      }
+      if (++receive_attempts > 100)
+        return make_error(sec::runtime_error,
+                          "too many unavailable_or_would_blocks");
+    } while (read_ret.index() != 0);
+    return none;
+  }
+
+  multiplexer_ptr mpx;
+  buffer_ptr shared_buf;
+  ip_endpoint ep;
+  udp_datagram_socket send_socket;
+  udp_datagram_socket recv_socket;
+};
+
+class dummy_application {
+  using buffer_type = std::vector<byte>;
+
+  using buffer_ptr = std::shared_ptr<buffer_type>;
+
+public:
+  dummy_application(buffer_ptr rec_buf)
```

Comment:
```suggestion
  explicit dummy_application(buffer_ptr rec_buf)
```

Even though it's "just" a unit test: please don't allow the compiler to turn any buffer it sees into an application. 😉